src/validation/validate_wasm.rs

```diff
@@ -0,0 +1,237 @@
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use anyhow::Result;
+use colored::Colorize;
+use impl_serde::serialize as serde_hex;
+use parity_wasm::elements::Module;
+
+/// Marker inserted by the ink! codegen for an error which can't
+/// be checked at compile time.
+const INK_ENFORCE_ERR: &str = "__ink_enforce_error_";
+
+/// Errors which may occur when forwarding a call is not allowed.
+///
+/// We insert markers for these errors in the generated contract code.
+/// This is necessary since we can't check these errors at compile time
+/// of the contract.
+/// `cargo-contract` checks the contract code for these error markers
+/// when building a contract and fails if it finds markers.
+#[derive(codec::Encode, codec::Decode)]
+pub enum EnforcedErrors {
+    /// The below error represents calling a `&mut self` message in a context that
+    /// only allows for `&self` messages. This may happen under certain circumstances
+    /// when ink! trait implementations are involved with long-hand calling notation.
+    #[codec(index = 1)]
+    CannotCallTraitMessage {
+        /// The trait that defines the called message.
+        trait_ident: String,
+        /// The name of the called message.
+        message_ident: String,
+        /// The selector of the called message.
+        message_selector: [u8; 4],
+        /// Is `true` if the `self` receiver of the ink! message is `&mut self`.
+        message_mut: bool,
+    },
+    /// The below error represents calling a constructor in a context that does
+    /// not allow calling it. This may happen when the constructor defined in a
+    /// trait is cross-called in another contract.
+    /// This is not allowed since the contract to which a call is forwarded must
+    /// already exist at the point when the call to it is made.
+    #[codec(index = 2)]
+    CannotCallTraitConstructor {
+        /// The trait that defines the called constructor.
+        trait_ident: String,
+        /// The name of the called constructor.
+        constructor_ident: String,
+        /// The selector of the called constructor.
+        constructor_selector: [u8; 4],
+    },
+}
+
+/// Validates the import section in the Wasm.
+///
+/// The checks currently fall into two categories:
+/// - Known bugs for which we want to recommend a solution.
+/// - Markers inserted by the ink! codegen for errors which can't be checked at compile time.
+pub fn validate_import_section(module: &Module) -> Result<()> {
+    let imports = module
+        .import_section()
+        .expect("import section must exist")
```

Comment:
Come to think about it I think the correct way of handling this is to actually have the validation return `Ok(())` if no import section is available ‒ since there are no invalid imports to be found.

I adapted this PR and added a test for this behavior.

Reply:
But shouldn't it import at least `seal_deploy` and `seal_call`, or is that checked somewhere else?

Reply:
`seal_deploy` and `seal_call` _exports_ need to be declared for a valid `contracts-pallet` Wasm smart contract.

Reply:
:facepalm:

Reply:
Hmm, does it make sense then to validate the Wasm for those exports and fail building the contract otherwise?

Reply:
Well, if those exports do not exist the Wasm is also invalid. Should normally not happen if you compile with ink! though.